test/AdaptiveSpinLockTest.cpp

```diff
@@ -62,27 +62,94 @@ TEST(AdaptiveSpinLockTest, WorksWithStdUniqueLock) {
   lock.unlock();
 }
 
-const int thread_num = 7;
-int val;
-pipert::AdaptiveSpinLock* g_lock = nullptr;
-
-void SharedIncrementor(int times) {
-  ASSERT_TRUE(g_lock);
-  for (int i = 0; i < times; i++) {
-    std::lock_guard<pipert::AdaptiveSpinLock> guard(*g_lock);
-    val++;
+namespace {
+
+class SharedIncrementor {
+ public:
+  void StartThreads(void func (SharedIncrementor&)) {
+    for(unsigned int i = 0; i < threads_.size(); ++i)
+      threads_[i] = std::thread(func, std::ref(*this));
+  }
+  void JoinThreads() {
+    for(unsigned int i = 0; i < threads_.size(); ++i)
+      threads_[i].join();
+  }
+
+  void SetThreads(const int count) { threads_.resize(count); }
+  int GetThreads() const { return threads_.size(); }
+  void SetCycles(const int cycles) { cycles_ = cycles; }
+  int GetCycles() const { return cycles_; }
+  void ResetVal() { val_ = 0; }
+  void IncrementVal() { val_++; }
+  int GetVal() const { return val_; }
+
+ private:
+  std::vector<std::thread> threads_;
+  int cycles_;
+  int val_;
+};
+
+static SharedIncrementor& SharedIncrementorInstance() {
+  static SharedIncrementor si;
```

Comment:
This is a bit suspicious. With this, you can create a ingleton object on demand, and leave it to the rest of the tests. So it will be created during one random test and will be reused for the others. This is dangeroud as if one tests screws this up, the rest can be affected. I would rather make sure to reinitialize this helpre object from scratch on every test start.

Reply:
It is a remnant from a time that the class was only used by a single test, and it made sense to have a static instance of the class. It definitely is a bad idea to leave tests to be responsible for perfectly resetting the class before running, and even then, that might not always be possible. It really is better to just reconstruct it for each test. Thanks for pointing that out.